pkgs/development/cuda-modules/setup-hooks/auto-fix-elf-files.sh

```diff
@@ -55,10 +62,51 @@ autoFixElfFiles() {
     elif elfHasDynamicSection "$f"; then
       # patchelf returns an error on statically linked ELF files, and in
       # practice fixing actions all involve patchelf
-      echo "autoFixElfFiles: using $fixAction to fix $f" >&2
+      (( "${NIX_DEBUG:-0}" >= 1 )) && echo "autoFixElfFiles: using $fixAction to fix $f" >&2
       $fixAction "$f"
     elif (( "${NIX_DEBUG:-0}" >= 1 )); then
       echo "autoFixElfFiles: skipping a statically-linked ELF file $f"
     fi
   done
 }
+
+elfBuildRunpathStrings() {
+    local path
+
+    for path in "${elfPrependRunpaths[@]}" ; do
+        elfAddRunpathsPrefix="$elfAddRunpathsPrefix:$path"
+    done
+    elfAddRunpathsPrefix="${elfAddRunpathsPrefix#:}"
+
+    for path in "${elfAddRunpaths[@]}" ; do
+        elfAddRunpathsSuffix="$elfAddRunpathsSuffix:$path"
+    done
+    elfAddRunpathsSuffix="${elfAddRunpathsSuffix#:}"
+}
+
+# Expects that elfAddRunpathPrefix and elfAddRunpathSuffix are set
+elfAddRunpathsAction() {
+    local origPath="$(patchelf --print-rpath "$1")"
+    local newPath
+
+    newPath="$elfAddRunpathsPrefix"
+    newPath="${newPath}${newPath:+:}${origPath}"
+    newPath="${newPath}${elfAddRunpathSuffix:+:}${elfAddRunpathSuffix}"
+
+    (( "${NIX_DEBUG:-0}" >= 4 )) && echo patchelf --set-rpath "$newPath" "$1" >&2
+    patchelf --set-rpath "$newPath" "$1"
+}
+
+elfAddRunpathsHook() {
+    [[ -z "${elfAddRunpaths[@]}" ]] && [[ -z "${elfPrependRunpaths[@]}" ]] && return
+
+    echo "Executing elfAddRunpaths: ${elfAddRunpaths[@]}" >&2
+    [[ -z "${elfPrependRunpaths[@]}" ]] || echo "elfPrependRunpaths: ${elfPrependRunpaths[@]}" >&2
+
+    declare elfAddRunpathsPrefix
+    declare elfAddRunpathsSuffix
+    elfBuildRunpathStrings
+    autoFixElfFiles elfAddRunpathsAction
```

Comment:
if `autoFixElfFiles` is only used there, I wonder if it's not simpler to inline it. As much as I liked my higher-order solution :stuck_out_tongue:  (or, another solution if we want to keep the code nicely modular in those different bash functions, would be to hardcode `elfAddRunpathsAction` instead of `fixAction`). Unless you envision `autoFixElfFiles` to be useful for other things?